### PR TITLE
Instrument spring-messaging to reduce chance of disconnected traces

### PR DESCRIPTION
**Pull request by mcculls**
# What Does This Do

Captures a new span around `spring-messaging` annotated listeners, such as `@SqsListener`

The resource name of this span is `<simple_class_name>.<method_name>`

If there's an active trace context then that is assumed to be the appropriate parent context for this message span. However if there is no active trace context the tracer attempts to (re)extract any distributed trace context from the message.

# Motivation

This should reduce chance of disconnected traces if the extracted context becomes lost, for example during asynchronous processing, but the original context is still available in the message.

# Additional Notes

You can use this JVM option to disable this new feature:
```
-Ddd.integration.spring-messaging.enabled=false
```
or this environment variable:
```
DD_INTEGRATION_SPRING_MESSAGING_ENABLED=false
```

# Example traces

Previous behaviour could lead to the trace ending at `SQS.ReceiveMessage`:
Previous behaviour could lead to the trace ending at SQS.ReceiveMessage

Now the trace includes the call from `spring-messaging` to the listener:
Now the trace includes the call from spring-messaging to the listener

### CODE REVIEW

File: dd-java-agent/instrumentation/spring-messaging-4/src/test/groovy/SpringListenerSQSTest.groovy [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    then:
+    def sendingSpan
+    assertTraces(4, SORT_TRACES_BY_START) {
+      trace(3) {
+        sendMessage(it, address, span(2))
+        getQueueUrl(it, address, span(2))
+        basicSpan(it, "parent")
+        sendingSpan = span(0)
```
Comment (PerfectSlayer) [inline dd-java-agent/instrumentation/spring-messaging-4/src/test/groovy/SpringListenerSQSTest.groovy:44]: Nice trick!

File: dd-java-agent/instrumentation/spring-messaging-4/src/main/java/datadog/trace/instrumentation/springmessaging/SpringMessageHandlerInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import net.bytebuddy.asm.Advice;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.handler.invocation.InvocableHandlerMethod;
+
+@AutoService(Instrumenter.class)
+public final class SpringMessageHandlerInstrumentation extends Instrumenter.Tracing
+    implements Instrumenter.ForSingleType {
+
+  public SpringMessageHandlerInstrumentation() {
+    super("spring-messaging");
```
Comment (PerfectSlayer) [inline dd-java-agent/instrumentation/spring-messaging-4/src/main/java/datadog/trace/instrumentation/springmessaging/SpringMessageHandlerInstrumentation.java:27]: Is it worth adding `spring-messaging-4` as alias?

**Review by PerfectSlayer — APPROVED** [review]
LGTM

File: dd-java-agent/instrumentation/spring-messaging-4/src/main/java/datadog/trace/instrumentation/springmessaging/SpringMessageHandlerInstrumentation.java (hunk 2 above)
Comment (mcculls) [reply dd-java-agent/instrumentation/spring-messaging-4/src/main/java/datadog/trace/instrumentation/springmessaging/SpringMessageHandlerInstrumentation.java:27]: Is that something we're doing now to improve the telemetry data?

**Pull request merged**